retrofit/src/main/java/retrofit2/Retrofit.java

```diff
@@ -194,24 +205,47 @@ private void validateServiceInterface(Class<?> service) {
         if (!platform.isDefaultMethod(method)
             && !Modifier.isStatic(method.getModifiers())
             && !method.isSynthetic()) {
-          loadServiceMethod(service, method);
+          loadServiceMethod(method);
         }
       }
     }
   }
 
-  ServiceMethod<?> loadServiceMethod(Class<?> service, Method method) {
-    ServiceMethod<?> result = serviceMethodCache.get(method);
-    if (result != null) return result;
+  ServiceMethod<?> loadServiceMethod(Method method) {
+    // Note: Once we are minSdk 24 this whole method can be replaced by computeIfAbsent.
+    Object lookup = serviceMethodCache.get(method);
+
+    if (lookup instanceof ServiceMethod<?>) {
+      // Happy path: method is already parsed into the model.
+      return (ServiceMethod<?>) lookup;
+    }
 
-    synchronized (service) {
-      result = serviceMethodCache.get(method);
-      if (result == null) {
-        result = ServiceMethod.parseAnnotations(this, method);
-        serviceMethodCache.put(method, result);
+    if (lookup == null) {
+      // Map does not contain any value. Try to put in a lock for this method. We MUST synchronize
+      // on the lock before it is visible to others via the map to signal we are doing the work.
+      Object lock = new Object();
+      synchronized (lock) {
+        lookup = serviceMethodCache.putIfAbsent(method, lock);
+        if (lookup == null) {
+          // On successful lock insertion, perform the work and update the map before releasing.
+          // Other threads may be waiting on lock now and will expect the parsed model.
+          ServiceMethod<Object> result = ServiceMethod.parseAnnotations(this, method);
+          serviceMethodCache.put(method, result);
+          return result;
+        }
+      }
+      if (lookup instanceof ServiceMethod<?>) {
```

Comment:
Take it or leave it: you could combine this case with the one on 218 and save 3 lines of code

Reply:
Yielding code roughly like this:

```
Object lookup = ...

if (lookup == null) {
  // Map does not contain any value. ...
  ...
}

if (lookup instanceof ServiceMethod) {
  // Happy path or happy path after putIfAbsent
  ...
}

synchronized (lookup) {

}
```

Reply:
Yeah. I also thought about dropping this check entirely. If you got back a `ServiceMethod` then it's harmless to lock on it and do another map lookup (for the same object). This is such a rare case, after all.

Reply:
I just dropped the second `instanceof` check. I wanted to keep the `instanceof` path first since this is called on _every_ method invocation. Trying to keep the happy path as speedy hot as possible, even if it only amounts to a tiny amount.